V3.0 working - Updates

mattburch opened the pull request:
Some of the default I/O STDERR messages made it difficult to thread multiple background executions of the tool. With the I/O adjustments, it is very easy to monitor the state of multiple requests.

```
$ ./gobuster dir -w /opt/lists/directory-list-lowercase-2.3-medium.txt -e -z -t 5 -k -u http://10.10.10.10/ 2>/dev/null
===============================================================
Gobuster v3.0.0 (beta)
by OJ Reeves (@TheColonial) & Christian Mehlmauer (@_FireFart_)
===============================================================
[+] Url:            http://10.10.10.10/
[+] Threads:        5
[+] Wordlist:       /opt/lists/directory-list-lowercase-2.3-medium.txt
[+] Status codes:   200,204,301,302,307,401,403
[+] User Agent:     gobuster/3.0.0 (beta)
[+] Expanded:       true
[+] Timeout:        10s
===============================================================
2019-05-24 09:30:42 Starting gobuster
===============================================================
[*] Wildcard Detected: http://10.10.10.10/ => 302
```

FireFart commented:
Hi,
what's wrong with log.Println? Both should print to stdout?

mattburch commented:
If multiple background processes are executed, having time stamps is
valuable to obtain the information. Println is an option as well but you
loose variable injection in the content.

On Fri, May 24, 2019 at 9:54 AM Christian Mehlmauer <
notifications@github.com> wrote:

> Hi,
> what's wrong with log.Println? Both should print to stdout?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OJ/gobuster/pull/146?email_source=notifications&email_token=ABKWVP6UZNTZEHKHSZIWQNDPW7XS7A5CNFSM4HPPSO4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWFNW7Y#issuecomment-495639423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWVPZMXPZ5QCR6IOHVXW3PW7XS7ANCNFSM4HPPSO4A>
> .
>

mattburch commented:
Typically I would execute the tool within a screen session and have
multiple threads executed against various hosts. Having stdout responses,
that are timestamped, allows me to review the session log and identify
start/stops and the status responses.

On Fri, May 24, 2019 at 9:54 AM Christian Mehlmauer <
notifications@github.com> wrote:

> Hi,
> what's wrong with log.Println? Both should print to stdout?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OJ/gobuster/pull/146?email_source=notifications&email_token=ABKWVP6UZNTZEHKHSZIWQNDPW7XS7A5CNFSM4HPPSO4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWFNW7Y#issuecomment-495639423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWVPZMXPZ5QCR6IOHVXW3PW7XS7ANCNFSM4HPPSO4A>
> .
>

FireFart commented:
I don't get the point sorry, log.Println already outputs the current timestamp on my machine?

mattburch commented:
The tool also can produce a lot of error stderr response content due to
standard net.Dial requests. The wildcard indicator is important, similar to
start/stop notifications and can get lost in 2>/dev/null for background
process execution. I agree stderr should be used for error conditions, but
WildCard flags are an app defined construct not a typical error condition.
Splitting the content into a stdout status condition along with an stderr
response allows for the conditional state of tool execution to be monitored
when error content is sent to /dev/null.

On Fri, May 24, 2019 at 10:05 AM Christian Mehlmauer <
notifications@github.com> wrote:

> I don't get the point sorry, log.Println already outputs the current
> timestamp on my machine?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OJ/gobuster/pull/146?email_source=notifications&email_token=ABKWVPZDAYQXKFICSGTOWZDPW7Y2JA5CNFSM4HPPSO4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWFO3GA#issuecomment-495644056>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWVP5KBZDZ2IN625BOULDPW7Y2JANCNFSM4HPPSO4A>
> .
>

FireFart commented:
The wildcard responses will be implemented again, but you also changed the log.Println to a fmt.Println with a timestamp which should not be necessary as it's already there

mattburch commented:
Start/Stop notifications were sent to stderr via a log statement. I
modified these to fmt.Print so they would be issued against stdout, but
that lost the timestamp which was a history tracker of when the tool
started and stopped. This tool is leveraged for pentesting efforts and some
clients require and indication when activities start/stop; which is not
possible to produce without a timestamp. Hence the stdout function of
providing a timestamp to stdout message content.

The Wildcard changes were submitted with a different git init comment.

On Fri, May 24, 2019 at 10:19 AM Christian Mehlmauer <
notifications@github.com> wrote:

> The wildcard responses will be implemented again, but you also changed the
> log.Println to a fmt.Println with a timestamp which should not be necessary
> as it's already there
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OJ/gobuster/pull/146?email_source=notifications&email_token=ABKWVP5SFQEI46OY7KJ2OODPW72OHA5CNFSM4HPPSO4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWFQJRQ#issuecomment-495649990>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWVP5RY45OLAVJHXQ4TSLPW72OHANCNFSM4HPPSO4A>
> .
>

mattburch commented:
Based on the original log.Print statement, it appeared applying a timestamp
was intended functionality. FMT does not produce the same interface logging
method, hence the function creation within libgobuster help file.

On Fri, May 24, 2019 at 10:23 AM Matt Burch <mburch44@gmail.com> wrote:

> Start/Stop notifications were sent to stderr via a log statement. I
> modified these to fmt.Print so they would be issued against stdout, but
> that lost the timestamp which was a history tracker of when the tool
> started and stopped. This tool is leveraged for pentesting efforts and some
> clients require and indication when activities start/stop; which is not
> possible to produce without a timestamp. Hence the stdout function of
> providing a timestamp to stdout message content.
>
> The Wildcard changes were submitted with a different git init comment.
>
> On Fri, May 24, 2019 at 10:19 AM Christian Mehlmauer <
> notifications@github.com> wrote:
>
>> The wildcard responses will be implemented again, but you also changed
>> the log.Println to a fmt.Println with a timestamp which should not be
>> necessary as it's already there
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/OJ/gobuster/pull/146?email_source=notifications&email_token=ABKWVP5SFQEI46OY7KJ2OODPW72OHA5CNFSM4HPPSO4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWFQJRQ#issuecomment-495649990>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABKWVP5RY45OLAVJHXQ4TSLPW72OHANCNFSM4HPPSO4A>
>> .
>>
>

FireFart commented:
So it's all about stderr I see. The fix for this would be another one, I will take a look at it thanks

FireFart commented:
@mattburch I added the features to the v3 branch. Would you mind giving it a test and if this works out for you?

mattburch commented:
So couple of notes. When the tool finishes:
```
===============================================================
2019/05/24 14:44:07 Finished
===============================================================
```

I modified this response to look like
```
===============================================================
2019/05/24 14:44:07 Finished - <URL|Domain>
===============================================================
```

This output provides more relevance to what task had actually completed if
this is executed as a background job.

Also the WildCard error output, below, still follows the stderr I/O channel
and can get lost if filtering error content.
```
Error: the server returns the same status code for every request.
http://10.10.10.10/0f4a110b-9953-402e-85e3-27dcb6f792ff => 302. To force
processing of Wildcard responses, specify the '--wildcard' switch
```

On Fri, May 24, 2019 at 2:02 PM Christian Mehlmauer <
notifications@github.com> wrote:

> @mattburch <https://github.com/mattburch> I added the features to the v3
> branch. Would you mind giving it a test and if this works out for you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OJ/gobuster/pull/146?email_source=notifications&email_token=ABKWVPZJUNIF2TKB4HSCBKDPXAUTBA5CNFSM4HPPSO4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWGE3LI#issuecomment-495734189>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWVP6QQZQSL4CQG5ZDS7TPXAUTBANCNFSM4HPPSO4A>
> .
>

FireFart commented:
Personally I would redirect each domain in a seperate file to seperate the domains. I think putting it in the finished message will distract normal users only scanning one domain / url.

For the wildcard message: This is an error and the program exits afterwards so this will always be in stderr, sorry

mattburch commented:
I understand the suggestion about piping the contents into a file, however
it appears this does not actually resolve the issue. Output logging of the
command execution does not capture start/stop times or input parameters.
This makes it difficult to present an accurate run book of activities
performed during an assessment exercise. This also leads back into the
previously mentioned item, that if your command is used to thread multiple
tasks, there isn't functionality to association the completion of one task
over the other.

In regards to the wildcard discussion, I understand the notification state
which initiates a programmatic closure of the application to be considered
an error response (stderr), however I don't see why you would be opposed to
following the metasploit status notification indicators [+] [-] [*] that a
possible wildcard condition was detected "[*] WildCard response identified".

On Fri, May 24, 2019 at 2:53 PM Christian Mehlmauer <
notifications@github.com> wrote:

> Personally I would redirect each domain in a seperate file to seperate the
> domains. I think putting it in the finished message will distract normal
> users only scanning one domain / url.
>
> For the wildcard message: This is an error and the program exits
> afterwards so this will always be in stderr, sorry
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OJ/gobuster/pull/146?email_source=notifications&email_token=ABKWVP3PEHODPBY34TM6MVLPXA2QZA5CNFSM4HPPSO4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWGJRFI#issuecomment-495753365>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWVP7TZ576YGCY6DABIHDPXA2QZANCNFSM4HPPSO4A>
> .
>

FireFart commented:
We can't fix everything in the tool, it should be kept simple and you can build wrappers around it to call it automatically. We also try not to do any wordlist mangling and stuff like this because it should be done outside of the tool. With v3 you can even use and import libgobuster in your own code and make use of it's functionality and call it many times in a loop.
The start and stop times are already printed out so there is a clear indication when the command started and when it finished. For the input parameters you could also log them in the wrapper.
The `--wildcard` is also clearly an error as the program exits when this happens and does not continue. If you provide the wildcard switch to the programm this message will never show up.
